File: go/testio.go
package main

import (
	"bytes"
	"encoding/hex"
	"fmt"
	"io"
	"os"
)

func processData(source io.Reader, destination io.Writer) error {
	const BUF_SIZE int = 8192
	var buffer [BUF_SIZE]byte
	var bytesRead int
	var err error
	for {
		totalBytesRead := 0
		for totalBytesRead < BUF_SIZE {
			bytesRead, err = source.Read(buffer[totalBytesRead:])
			if bytesRead == 0 {
				if totalBytesRead > 0 {
					for i := range totalBytesRead {
						buffer[i] ^= 0x5a
					}
					_, err = destination.Write(buffer[:totalBytesRead])
					if err != nil {
						return err
					}
					return nil
				}
			}
			totalBytesRead += bytesRead
		}
		for i := range BUF_SIZE {
			buffer[i] ^= 0x5a
		}
		_, err = destination.Write(buffer[:])
		if err != nil {
			return err
		}
	}

	// If short reads are okay:
	// for {
	// 	bytesRead, err := source.Read(buffer[:])
	// 	if bytesRead == 0 {
	// 		break
	// 	}
	// 	if err != nil {
	// 		return err
	// 	}
	// 	for i := range bytesRead {
	// 		buffer[i] ^= 0x5a
	// 	}
	// 	_, err = destination.Write(buffer[:bytesRead])
	// 	if err != nil {
	// 		return err
	// 	}
	// }
	// return nil
}

func main() {
	src, err := os.Open("input.bin")
	if err != nil {
		panic(err)
	}
	dst, err := os.Create("output.bin")
	if err != nil {
		panic(err)
	}
	defer src.Close()
	defer dst.Close()
	err = processData(src, dst)
	if err != nil {
		panic(err)
	}
	fmt.Println("File written to successfully.")

	in_memory := []byte("Hello this is some random text!")
	byteInput := bytes.NewReader(in_memory)
	var byteOutput bytes.Buffer
	err = processData(byteInput, &byteOutput)
	if err != nil {
		panic(err)
	}
	fmt.Println("Bytes copied successfully.")
	result := byteOutput.Bytes()
	fmt.Println(hex.EncodeToString(result))

	// net.Conn already implements io.Reader and io.Writer
}

File: testio.cpp
#include <algorithm>
#include <iostream>
#include <fstream>
#include <sstream>

void process_data(std::istream& source, std::ostream& destination) {
    const std::streamsize BUF_SIZE = 8192;
    char buffer[BUF_SIZE];
    std::streamsize bytes_read;
    while (true) {
        std::streamsize total_bytes_read = 0;
        while (total_bytes_read < BUF_SIZE) {
            source.read(buffer, BUF_SIZE);
            if (!source && !source.eof()) {
                throw std::runtime_error("Read error");
            }
            bytes_read = source.gcount();
            if (bytes_read == 0) {
                if (total_bytes_read > 0) {
                    for (std::streamsize i = 0; i < total_bytes_read; i++) {
                        buffer[i] ^= 0x5a;
                    }
                    destination.write(buffer, total_bytes_read);
                    if (!destination) {
                        throw std::runtime_error("Write error");
                    }
                    destination.flush();
                }
                return;
            }
            total_bytes_read += bytes_read;
        }
        for (std::streamsize i = 0; i < BUF_SIZE; i++) {
            buffer[i] ^= 0x5a;
        }
        destination.write(buffer, BUF_SIZE);
        if (!destination) {
            throw std::runtime_error("Write error");
        }
        destination.flush();
    }

    // If short reads are okay:
    // while (true) {
    //     source.read(buffer, BUF_SIZE);
    //     if (!source && !source.eof()) {
    //         throw std::runtime_error("Read error");
    //     }
    //     bytes_read = source.gcount();
    //     if (bytes_read == 0) {
    //         break;
    //     }
    //     std::transform(std::begin(buffer), std::begin(buffer) + bytes_read, std::begin(buffer), [](char c) { return c ^ 0x5a; });
    //     destination.write(buffer, bytes_read);
    //     if (!destination) {
    //         throw std::runtime_error("Write error");
    //     }
    //     destination.flush();
    // }
}

int main(void) {
    std::ifstream src("input.bin", std::ios::binary);
    if (!src) {
        std::cerr << "Error opening input file" << std::endl;
        return 1;
    }
    std::ofstream dst("output.bin", std::ios::binary);
    if (!dst) {
        std::cerr << "Error opening output file" << std::endl;
        return 1;
    }
    process_data(src, dst);
    std::cout << "File written to successfully." << std::endl;

    std::istringstream byte_input("Hello this is some random text!");
    std::ostringstream byte_output;
    process_data(byte_input, byte_output);
    std::cout << "Bytes copied successfully." << std::endl;
    std::string result = byte_output.str();
    std::for_each(result.begin(), result.end(), [](char c) { std::cout << std::hex << (int)c; });
    std::cout << "\n";

    // No straightforward built-in way to convert a socket descriptor to std::istream or std::ostream
    // However Boost.Asio offers socket stream wrappers https://www.boost.org/doc/libs/latest/doc/html/boost_asio.html
}
